src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt

```diff
@@ -776,4 +712,104 @@ class LicenceService(
+    licenceCopy.createdBy = createdBy
+    licenceCopy.mailingList.add(licenceCopy.responsibleCom!!)
+    licenceCopy.mailingList.add(createdBy!!)
+    licenceCopy.version = licencePolicyService.currentPolicy().version
+
+    if (newStatus == VARIATION_IN_PROGRESS) {
+      licenceCopy.variationOfId = licence.id
+    } else if (newStatus == IN_PROGRESS) {
+      licenceCopy.versionOfId = licence.id
+    }
+
+    val newLicence = licenceRepository.save(licenceCopy)
+
+    val standardConditions = licence.standardConditions.map {
+      it.copy(id = -1, licence = newLicence)
+    }
+
+    val bespokeConditions = licence.bespokeConditions.map {
+      it.copy(id = -1, licence = newLicence)
+    }
+
+    standardConditionRepository.saveAll(standardConditions)
+    bespokeConditionRepository.saveAll(bespokeConditions)
+
+    val additionalConditions = licence.additionalConditions.map {
+      val additionalConditionData = it.additionalConditionData.map { data ->
+        data.copy(id = -1)
+      }
+      val additionalConditionUploadSummary = it.additionalConditionUploadSummary.map { upload ->
+        upload.copy(id = -1)
+      }
+      it.copy(
+        id = -1,
+        licence = newLicence,
+        additionalConditionData = additionalConditionData,
+        additionalConditionUploadSummary = additionalConditionUploadSummary
+      )
+    }
+
+    var newAdditionalConditions = additionalConditionRepository.saveAll(additionalConditions).toMutableList()
+
+    newAdditionalConditions = newAdditionalConditions.map { condition ->
+      val updatedAdditionalConditionData = condition.additionalConditionData.map {
+        it.copy(additionalCondition = condition)
+      }
+
+      val updatedAdditionalConditionUploadSummary = condition.additionalConditionUploadSummary.map {
+        var uploadDetail = additionalConditionUploadDetailRepository.getReferenceById(it.uploadDetailId)
+        uploadDetail = uploadDetail.copy(id = -1, licenceId = newLicence.id, additionalConditionId = condition.id)
+        uploadDetail = additionalConditionUploadDetailRepository.save(uploadDetail)
+        it.copy(additionalCondition = condition, uploadDetailId = uploadDetail.id)
+      }
+
+      condition.copy(
+        additionalConditionData = updatedAdditionalConditionData,
+        additionalConditionUploadSummary = updatedAdditionalConditionUploadSummary
+      )
+    } as MutableList<AdditionalCondition>
+
+    additionalConditionRepository.saveAll(newAdditionalConditions)
+
+    val licenceEventMessage = when (newStatus) {
+      VARIATION_IN_PROGRESS -> "A variation was created for ${newLicence.forename} ${newLicence.surname} from ID ${licence.id}"
+      IN_PROGRESS -> "A new licence version was created for ${newLicence.forename} ${newLicence.surname} from ID ${licence.id}"
+      else -> "A licence copy with status ${newLicence.statusCode} was created for ${newLicence.forename} ${newLicence.surname} from ID ${licence.id}"
```

Comment:
wondering if it's worth having a precondition to check status is one of these, and throw illegals states here...

Reply:
yes done, if someone uses this for a new status it will make them create a new message for it,